### ISSUE TITLE
OpenFace FaceLandmarkVid Performance/Optimization

### ISSUE THREAD

**Issue by kaisark** [TadasBaltrusaitis/OpenFace#490]
Hi. While the results from the video were impressive, the program was running at a reported 8-10 FPS on the Jetson TX1 (ARM64 - Ubuntu 16). Is there any way to speed things up? I know that the DLIB facial detector can be slow, though there were recent improvements to speed things up. I also recall in python, for example, one can pass an argument to the detector that increases/decreases accuracy which can help speed things up. (e.g. rects = detector(frame, 0)). Also, the program can probably run every other frame for applications like pose/gaze detection...

Has anyone tackled OpenFace performance for real time applications?

-------------------------
(cvdlibpy35) nvidia@tegra-ubuntu:\~/cviz/OpenFace/build/bin$ ./FaceLandmarkVid -f ../../samples/driving-drowsy.mp4 
Reading the landmark detector/tracker from: model/main_ceclm_general.txt
Reading the landmark detector module from: model/cen_general.txt
Reading the PDM module from: model/pdms/In-the-wild_aligned_PDM_68.txt....Done
Reading the Triangulations module from: model/tris_68.txt....Done
Reading the intensity CEN patch experts from: model/patch_experts/cen_patches_0.25_of.dat....Done
Reading the intensity CEN patch experts from: model/patch_experts/cen_patches_0.35_of.dat....Done
Reading the intensity CEN patch experts from: model/patch_experts/cen_patches_0.50_of.dat....Done
Reading the intensity CEN patch experts from: model/patch_experts/cen_patches_1.00_of.dat....Done
Reading part based module....left_eye_28
Reading the landmark detector/tracker from: model/model_eye/main_clnf_synth_left.txt
Reading the landmark detector module from: model/model_eye/clnf_left_synth.txt
Reading the PDM module from: model/model_eye/pdms/pdm_28_l_eye_3D_closed.txt....Done
Reading the intensity CCNF patch experts from: model/model_eye/patch_experts/left_ccnf_patches_1.00_synth_lid_.txt....Done
Reading the intensity CCNF patch experts from: model/model_eye/patch_experts/left_ccnf_patches_1.50_synth_lid_.txt....Done
Done
Reading part based module....right_eye_28
Reading the landmark detector/tracker from: model/model_eye/main_clnf_synth_right.txt
Reading the landmark detector module from: model/model_eye/clnf_right_synth.txt
Reading the PDM module from: model/model_eye/pdms/pdm_28_eye_3D_closed.txt....Done
Reading the intensity CCNF patch experts from: model/model_eye/patch_experts/ccnf_patches_1.00_synth_lid_.txt....Done
Reading the intensity CCNF patch experts from: model/model_eye/patch_experts/ccnf_patches_1.50_synth_lid_.txt....Done
Done
Reading the landmark validation module....Done
Attempting to read from file: ../../samples/driving-drowsy.mp4
Device or file opened
Starting tracking
Reading the MTCNN face detector from: model/mtcnn_detector/MTCNN_detector.txt
Reading the PNet module from: model/mtcnn_detector/PNet.dat
Reading the RNet module from: model/mtcnn_detector/RNet.dat
Reading the ONet module from: model/mtcnn_detector/ONet.dat

![computervision gif-downsized_large](https://user-images.githubusercontent.com/3580312/41828816-62e77844-7805-11e8-8273-dc511ed8b56d.gif)

**Comment by TadasBaltrusaitis** [comment]
Hi,

Great to hear that you managed to get it running on an ARM64 system. A couple of comments on performance:
1. actually the default face detector used is not dlib, but my implementation of an MTCNN face detector, which is comparable in speed but is much more accurate (although I don't know how fast it is on a Jetson TX1), it would be worth profiling that
2.  All of the deep neural net stuff in OpenFace runs on CPU (the CEN patch experts, MTCNN face detector, and a CNN tracking validation), moving it to GPU should speed things up considerably. One of the reasons I have not done so already is to keep compatibility with CPU only and lower end machines, but I am thinking of adding GPU support, just not anytime soon.
3. The most expensive step is landmark detection, and that needs to run every frame for good results
4. Not all landmarks need to be detected at every scale of processing, I am actually currently working on implementing this, and this should give a considerable boost in speed, but you will have to wait 4-5 weeks for it to get integrated into the master release of OpenFace
5. In general OpenFace is suitable for real time processing on a number of machines (it runs close to 30fps on my Surface Laptop)

P.S. it would be great if you could write up what you needed to change to get it running! I'm sure a lot of people would find it very helpful to know how to get OpenFace running on ARM

Thanks,
Tadas

**Comment by lzwhard** [comment]
I have runned on arm too, and the big problem is slow. 
My exploring to speed up:
      --  remove the eye refinement, the fps will double

**Comment by guihui123** [comment]
> I have runned on arm too, and the big problem is slow.
> My exploring to speed up:
> -- remove the eye refinement, the fps will double

Excuse me,how can i remove the eye refinement, you mean GazeAnalysis::EstimateGaze?

**Comment by lyyiangang** [comment]
@TadasBaltrusaitis  hi sir, Could you provide models that can use GPU? e.g. tensorflow model or pytorch model?

**Comment by TadasBaltrusaitis** [comment]
OpenFace is composed of a number of modules and trained models, only one of the modules was trained with a standard deep learning framework (CEN patch experts). However, all of them are running on CPU without GPU support by default, so you would need to implement GPU optimization yourself.

Thanks,
Tadas

**Comment by noamzafrir** [comment]
> I have runned on arm too, and the big problem is slow.
> My exploring to speed up:
> -- remove the eye refinement, the fps will double

Hi,
What is Eye Refinement and how can it be removed?

**Comment by TadasBaltrusaitis** [comment]
Maybe it is referring to the following:
https://github.com/TadasBaltrusaitis/OpenFace/issues/882

**Comment by futureisatyourhand** [comment]
> Hi,
> 
> Great to hear that you managed to get it running on an ARM64 system. A couple of comments on performance:
> 
> 1. actually the default face detector used is not dlib, but my implementation of an MTCNN face detector, which is comparable in speed but is much more accurate (although I don't know how fast it is on a Jetson TX1), it would be worth profiling that
> 2. All of the deep neural net stuff in OpenFace runs on CPU (the CEN patch experts, MTCNN face detector, and a CNN tracking validation), moving it to GPU should speed things up considerably. One of the reasons I have not done so already is to keep compatibility with CPU only and lower end machines, but I am thinking of adding GPU support, just not anytime soon.
> 3. The most expensive step is landmark detection, and that needs to run every frame for good results
> 4. Not all landmarks need to be detected at every scale of processing, I am actually currently working on implementing this, and this should give a considerable boost in speed, but you will have to wait 4-5 weeks for it to get integrated into the master release of OpenFace
> 5. In general OpenFace is suitable for real time processing on a number of machines (it runs close to 30fps on my Surface Laptop)
> 
> P.S. it would be great if you could write up what you needed to change to get it running! I'm sure a lot of people would find it very helpful to know how to get OpenFace running on ARM
> 
> Thanks,
> Tadas

hello,do you complete the model running on the gpu?
thanks
